packages/protocol/contracts/layer1/based/TaikoInbox.sol

```diff
@@ -176,33 +175,32 @@ abstract contract TaikoInbox is EssentialContract, ITaikoInbox, IProposeBatch, I
                     blobByteOffset: params.blobParams.byteOffset,
                     blobByteSize: params.blobParams.byteSize,
                     gasLimit: config.blockMaxGasLimit,
-                    lastBlockId: 0, // to be initialised later
+                    lastBlockId: lastBatch.lastBlockId + uint64(params.blocks.length),
                     lastBlockTimestamp: lastBlockTimestamp,
-                    //
                     // Data for the L2 anchor transaction, shared by all blocks in the batch
                     anchorBlockId: anchorBlockId,
                     anchorBlockHash: blockhash(anchorBlockId),
                     baseFeeConfig: config.baseFeeConfig
                 });
 
-                uint64 nBlocks = uint64(params.blocks.length);
-
                 require(info_.anchorBlockHash != 0, ZeroAnchorBlockHash());
 
-                info_.lastBlockId = stats2.numBatches == config.forkHeights.pacaya
-                    ? stats2.numBatches + nBlocks - 1
-                    : lastBatch.lastBlockId + nBlocks;
-
                 bytes32 txListHash = keccak256(_txList);
                 (info_.txsHash, info_.blobHashes) = _calculateTxsHash(txListHash, params.blobParams);
 
                 meta_ = BatchMetadata({
                     infoHash: keccak256(abi.encode(info_)),
                     prover: info_.proposer,
                     batchId: stats2.numBatches,
-                    proposedAt: uint64(block.timestamp)
+                    proposedAt: uint64(block.timestamp),
+                    firstBlockId: lastBatch.lastBlockId + 1
                 });
 
+                // Fork activation shall be L2 block based
+                // This check means first block of this batch is at least equals to shasta block
+                // height
+                require(meta_.firstBlockId >= config.forkHeights.shasta, ForkNotActivated());
```

Comment:
Lets also prepare for the fork after shasta by checking:

require(config.forkHeights.unzen == 0 || info_.lastBlockId < config.forkHeights.unzenm)